core/src/avm2/activation.rs

```diff
@@ -1467,16 +1467,32 @@ impl<'a, 'gc> Activation<'a, 'gc> {
             // rather than it's string representation.
 
             let name_value = self.context.avm2.peek(0);
-            let object = self.context.avm2.peek(1);
-            if !name_value.is_primitive() {
-                let object = object.coerce_to_object_or_typeerror(self, None)?;
-                if let Some(dictionary) = object.as_dictionary_object() {
-                    let _ = self.pop_stack();
-                    let _ = self.pop_stack();
-                    let value = dictionary.get_property_by_object(name_value.as_object().unwrap());
-                    self.push_stack(value);
-
-                    return Ok(FrameControl::Continue);
+            let object_value = self.context.avm2.peek(1);
+
+            if let Value::Object(object) = object_value {
+                match name_value {
+                    Value::Object(name_object) => {
+                        if let Some(dictionary) = object.as_dictionary_object() {
+                            let _ = self.pop_stack();
+                            let _ = self.pop_stack();
+                            let value = dictionary.get_property_by_object(name_object);
+                            self.push_stack(value);
+
+                            return Ok(FrameControl::Continue);
+                        }
+                    }
+                    Value::Integer(name_int) if name_int >= 0 => {
```

Comment:
nit: I'd put this branch first, with a comment explaining that this is just a fast-path, with identical behavior intended.